src/well_known_messages.cc

```diff
@@ -0,0 +1,248 @@
+        {
+            auto elementCount = rows * columns;
+            NumericArrayResize(GetTypeCodeForSize(sizeof(double)), 2, start, elementCount);
+            auto array = *(LV2DArrayHandle*)start;
+            (*array)->dimensionSizes[0] = rows;
+            (*array)->dimensionSizes[1] = columns;
+            // Protect against a malformed message where the amount of data sent doesn't match the dimension sizes.
+            // LV will automatically pad/handle writing less data than was allocated to the array so we just need
+            // to make sure we don't write more data than was allocated to the array.
+            auto dataValue = std::static_pointer_cast<const LVRepeatedMessageValue<double>>(dataFieldValue);
+            auto lvByteCount = elementCount * sizeof(double);
+            auto protobufByteCount = dataValue->_value.size() * sizeof(double);
+            memcpy((*array)->bytes<double>(), dataValue->_value.data(), std::min(lvByteCount, protobufByteCount));
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        void Double2DArray::CopyArrayFromClusterToMessage(int totalElements, LV2DArrayHandle array, const std::shared_ptr<LVMessageValue>& dataFieldValue)
+        {
+            auto arrayData = (*array)->bytes<double>();
+            auto dataValue = std::static_pointer_cast<LVRepeatedMessageValue<double>>(dataFieldValue);
+            dataValue->_value.Reserve(totalElements);
+            auto dest = dataValue->_value.AddNAlreadyReserved(totalElements);
+            memcpy(dest, arrayData, totalElements * sizeof(double));
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        const std::string& String2DArray::GetMessageName()
+        {
+            static const std::string messageName = "ni_protobuf_types_String2DArray";
+            return messageName;
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        const std::string& String2DArray::GetMessageUrl()
+        {
+            static const std::string messageUrl = "ni.protobuf.types.String2DArray";
+            return messageUrl;
+        }
+
+        //---------------------------------------------------------------------
+        //---------------------------------------------------------------------
+        void String2DArray::CopyArrayFromMessageToCluster(int rows, int columns, const std::shared_ptr<const LVMessageValue>& dataFieldValue, int8_t* start)
+        {
+            auto elementCount = rows * columns;
+            NumericArrayResize(GetTypeCodeForSize(sizeof(char*)), 2, start, elementCount);
```

Comment:
nit: the array doesn't actually contain char pointers
```suggestion
            NumericArrayResize(GetTypeCodeForSize(sizeof(LStrHandle)), 2, start, elementCount);
```

Reply:
Yeah, I originally had `void*` and then changed to `char*` with similar logic in `cluster_copier.cc`. I like use of `LStrHandle` better and have updated both locations.